docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+developer toil.
+
+Additionally, introducing `JobSet` as an infrastructure layer would allow us to introduce batch
+workload features such as
+[the PodFailurePolicy](https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)
+and [the PodDisruptionCondition](https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+easily.
+
+Please also see the [Kubernetes JobSet and Kubeflow Training Operator collaboration document](https://docs.google.com/document/d/1C2ev7yRbnMTlQWbQCfX7BCCHcLIAGW2MP9f7YeKl2Ck/edit?usp=sharing).
+
+### User Value
+
+In addition to the above motivation, we will address the following user feedback while implementation:
+
+- Confusion around Workers: https://github.com/kubeflow/training-operator/issues/1790
+- Support batch/v1 `Job` features: https://github.com/kubeflow/training-operator/issues/1718
+- ExitCodes for PodFailurePolicy: https://github.com/kubeflow/training-operator/issues/1749
+- Migrate to MPI V2 API: https://github.com/kubeflow/training-operator/issues/1906
+
+### Personas
+
+We can identify the following personas of Training Operator:
+
+1. **DevOps Engineer**. They are familiar with Kubernetes concepts and they know how to manage the
+   Kubernetes workloads. Usually, they are not experts in ML frameworks and ML algorithms.
+1. **MLOps Engineer**. They are familiar with ML frameworks and they know how to configure
+   distributed PyTorch settings such as rendezvous backends or MPI configuration. Usually, they are
+   not experts in Kubernetes and ML algorithms.
+1. **Data Scientists/ML Engineers**. They create model architectures and advanced ML algorithms to train models.
+   They prefer to use Python for their work. They are aware of `torch.nn` APIs, but not with
+   `torch.distributed` and Kubernetes concepts to scale model training.
+
+Based on the above personas, we should build an API that everyone will benefit from.
+
+### Goals
+
+- Introduce the `TrainingRuntime` and `ClusterTrainingRuntime` APIs that will store blueprints
+  for model training and LLM fine-tuning using various ML frameworks. These runtimes will be built
+  on top of `JobSet` APIs with additional functionality for special use-cases.
+  For example, training using MPI orchestration.
+- Introduce Kubeflow `TrainJob` API that allows to reuse these runtimes and quickly start a new
+  training job without understanding complex Kubernetes APIs.
+- Update Kubeflow Training SDK to allow data scientists quickly create and monitor `TrainJobs`.
+- Create community-supported `ClusterTrainingRuntime` for distributed training with PyTorch and MPI.
+- Create community-supported `ClusterTrainingRuntime` for LLM fine-tuning for various foundational
+  models (e.g. Mistral, LLama-70b, Gemma-7b).
+- Work on the following `JobSet` improvements: https://github.com/kubernetes-sigs/jobset/issues/463
```

Comment:
Additionally, we need to work on https://github.com/kubernetes-sigs/jobset/issues/570 so that we can netively support the Elastic PyTorchJob.